lib/path/default.nix

```diff
@@ -92,8 +106,82 @@ let
     # An empty string is not a valid relative path, so we need to return a `.` when we have no components
     (if components == [] then "." else concatStringsSep "/" components);
 
+
+  # Takes a Nix path value and deconstructs it into the filesystem root
+  # (generally `/`) and a subpath
+  deconstructPath =
+    let
+      go = components: path:
+        # If the parent of a path is the path itself, then it's a filesystem root
+        if path == dirOf path then { root = path; inherit components; }
+        else go ([ (baseNameOf path) ] ++ components) (dirOf path);
+    in go [];
```

Comment:
Components seems more widely used: https://github.com/NixOS/nixpkgs/pull/210423#discussion_r1072691860

Reply:
I think `go` as a convention is fine. It's a name that doesn't usually conflict with anything else and to the people familiar with it it's clear what the function is (a very local recursive function). It also doesn't make sense to invent a new name for it since it's just a helper function that does the same thing as the parent function. I don't think anything would be gained by naming this differently.

Particularly in this case if we wanted to call the main function `splitPath` and this local function `split`, then we already have a slight potential confusion between the two. Additionally however we also have `builtins.split` already imported into this scope, and while the `split` here would shadow it, it can still be confusing.

So I think `go` is just fine here

Reply:
Alright, `split` doesn't fit because it shadows an existing symbol. Still, `go` isn't helpful for reading the code, no matter how internal it is. It should at least hint at why it's there. So why not `recurse`?

Reply:
`recurse` is a bit less conventional, but yeah I'll change it